base/acme/src/org/dogtagpki/acme/validator/HTTP01Validator.java

```diff
@@ -0,0 +1,155 @@
+
+import java.net.URI;
+
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.Response.ResponseBuilder;
+import javax.ws.rs.core.UriInfo;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang.RandomStringUtils;
+import org.apache.http.HttpEntity;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.dogtagpki.acme.ACMEAuthorization;
+import org.dogtagpki.acme.ACMEChallenge;
+import org.dogtagpki.acme.ACMEError;
+import org.dogtagpki.acme.ACMEIdentifier;
+
+/**
+ * @author Endi S. Dewata
+ */
+public class HTTP01Validator extends ACMEValidator {
+
+    public static org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(HTTP01Validator.class);
+
+    public ACMEChallenge createChallenge(
+            UriInfo uriInfo,
+            String authzID,
+            String token) throws Exception {
+
+        // TODO: find better way to generate challenge ID
+
+        String challengeID = RandomStringUtils.randomAlphanumeric(10);
+        logger.info("Creating HTTP challenge: " + challengeID);
+
+        URI challengeURI = uriInfo.getBaseUriBuilder().path("chall").path(challengeID).build();
+
+        ACMEChallenge challenge = new ACMEChallenge();
+        challenge.setID(challengeID);
+        challenge.setAuthzID(authzID);
+        challenge.setType("http-01");
+        challenge.setURL(challengeURI);
+        challenge.setToken(token);
+
+        return challenge;
+    }
+
+    public void validateChallenge(
+            ACMEAuthorization authorization,
+            ACMEChallenge challenge) throws Exception {
+
+        // HTTP-01 key authorization
+        // https://tools.ietf.org/html/rfc8555
+
+        String accountID = authorization.getAccountID();
+        String token = challenge.getToken();
+        String keyAuthorization = token + "." + accountID;
+        logger.info("Key authorization: " + keyAuthorization);
+
+        ACMEIdentifier identifier = authorization.getIdentifier();
+        String hostname = identifier.getValue();
+        String validationPath = "/.well-known/acme-challenge/" + token;
+        URI validationURL = new URI("http", hostname, validationPath, null);
+
+        // TODO: make it configurable
+
+        int maxCount = 5;
+        int interval = 5;
+
+        int count = 1;
+        String value = null;
+        Exception exception = null;
+
+        while (true) {
+            try {
+                value = retrieveURL(validationURL);
+                break;
+            } catch (Exception e) {
+                logger.error(e.getMessage());
+                if (count >= maxCount) {
+                    exception = e;
+                    break;
+                }
+                Thread.sleep(interval * 1000);
```

Comment:
I think it is less of a concern here, but same comments as the DNS based challenge go here.

Reply:
See my comments for DNS challenge.